### PR TITLE
[SYSTEMDS-3685] Python FFT

### PR DESCRIPTION
**Pull request by fzoepffel**
## Overview
Python tests for the FFT and IFFT implementations were added. Also validation checks were added that should hopefully filter out all edge cases. Handling of zero matrix inputs and empty matrices was also implemented. This PR has been rebased on the main PR that implements FFT and IFFT and integrates it into DML. Thus it has to be merged after this PR.  See https://github.com/apache/systemds/pull/1983. 

## Testing
Run the tests in systemds/src/main/python/tests/matrix/test_fft.py and see if they are all going through. The tests can be run by doing the following: 
### Go to the root of the project. Then run:
```
python3.8 -m venv venv
source venv/bin/activate 
```

### make sure to set the following env variables in your venv:
```
export SYSTEMDS_ROOT=$(pwd)
export PATH=$SYSTEMDS_ROOT/bin:$PATH
export SYSDS_QUIET=1
```

### Then run:
```
pip install --upgrade pip
pip install wheel numpy py4j scipy scikit-learn requests pandas unittest-parallel
mvn clean package -P distribution    
cd src/main/python/
python create_python_dist.py
python -m unittest discover -s tests -p 'test_fft.py'
```

### For deactivating the venv run:
`deactivate`

After validating the tests review the code for the validation and check if any edge cases might be missing in systemds/src/main/java/org/apache/sysds/parser/BuiltinFunctionExpression.java and also check the handling of non-zero matrices in systemds/src/main/java/org/apache/sysds/runtime/matrix/data/LibCommonsMath.java. 

Contributors:
@fzoepffel
@jessicapriebe
@mufwan

### CODE REVIEW

**Comment by Baunsgaard** [comment]
Hi @fzoepffel 

Good to see the PR, I think you accidentally added your virtual environments to the commit,
you can add a the venv folder to .gitignore and remove it.

Thanks

**Comment by fzoepffel** [comment]
> Hi @fzoepffel
> 
> Good to see the PR, I think you accidentally added your virtual environments to the commit, you can add a the venv folder to .gitignore and remove it.
> 
> Thanks

Removed the venv folder and added it to gitignore

**Comment by Baunsgaard** [comment]
> > Hi @fzoepffel
> > Good to see the PR, I think you accidentally added your virtual environments to the commit, you can add a the venv folder to .gitignore and remove it.
> > Thanks
> 
> Removed the venv folder and added it to gitignore

there is still one other venv folder there.

**Comment by fzoepffel** [comment]
> > > Hi @fzoepffel
> > > Good to see the PR, I think you accidentally added your virtual environments to the commit, you can add a the venv folder to .gitignore and remove it.
> > > Thanks
> > 
> > 
> > Removed the venv folder and added it to gitignore
> 
> there is still one other venv folder there.

Two even :D Missed those ones. Now it should be correct

File: src/main/python/tests/matrix/test_fft.py
Comment (Baunsgaard) [inline src/main/python/tests/matrix/test_fft.py:1]: good tests

File: src/main/python/systemds/operator/nodes/matrix.py [hunk 1]
```diff
@@ -343,6 +345,32 @@ def cholesky(self, safe: bool = False) -> 'Matrix':
         """
         return Matrix(self.sds_context, 'cholesky', [self])
 
+    def fft(self):
+        """
+        Performs the Fast Fourier Transform (FFT) on this matrix.
+        Returns a Matrix object representing the FFT (complex numbers represented as two matrices).
```
Comment (Baunsgaard) [inline src/main/python/systemds/operator/nodes/matrix.py:351]: the syntax for documentation is not followed, take a look at some of the other examples.

File: src/main/java/org/apache/sysds/runtime/privacy/propagation/PrivacyPropagator.java
Comment (Baunsgaard) [inline src/main/java/org/apache/sysds/runtime/privacy/propagation/PrivacyPropagator.java:1]: This file should not be modified

**Review by Baunsgaard — COMMENTED** [review]
HI @fzoepffel 

good start, i like the contribution. But the PR makes many syntactic changes to files that should not be formated. Please revert the formatting of files, to only include necessary changes.

Thanks

File: src/main/python/systemds/operator/nodes/matrix.py (hunk 1 above)
Comment (fzoepffel) [reply src/main/python/systemds/operator/nodes/matrix.py]: this code is not even necessary. Had forgotten to remove it. Removed it now

**Comment by Baunsgaard** [comment]
This PR is currently being merged in #2009
